src/consensus.rs

```diff
@@ -822,92 +760,229 @@ fn handle_committed_entries(
-    let channel = make_grpc_channel(
-        bootstrap_timeout,
-        bootstrap_timeout,
-        bootstrap_uri,
-        tls_config,
-    )
-    .await
-    .map_err(|err| anyhow!("Failed to create timeout channel: {}", err))?;
-    let mut client = RaftClient::new(channel);
-    Ok(client
-        .who_is(tonic::Request::new(GrpcPeerId { id: peer_id }))
-        .await?
-        .into_inner()
-        .uri
-        .parse()?)
+    consensus_config: Arc<ConsensusConfig>,
+    consensus_state: ConsensusStateRef,
+    transport_channel_pool: Arc<TransportChannelPool>,
+}
+
+impl RaftMessageBroker {
+    pub fn new(
+        runtime: Handle,
+        bootstrap_uri: Option<Uri>,
+        tls_config: Option<ClientTlsConfig>,
+        consensus_config: ConsensusConfig,
+        consensus_state: ConsensusStateRef,
+        transport_channel_pool: Arc<TransportChannelPool>,
+    ) -> Self {
+        Self {
+            senders: HashMap::new(),
+            runtime,
+            bootstrap_uri,
+            tls_config,
+            consensus_config: consensus_config.into(),
+            consensus_state,
+            transport_channel_pool,
+        }
+    }
+
+    pub fn send(&mut self, messages: impl IntoIterator<Item = RaftMessage>) {
+        for message in messages {
+            let peer_id = message.to;
+
+            let sender = match self.senders.get(&peer_id) {
+                Some(sender) => sender,
+                None => {
+                    log::debug!("Spawning message sender task for peer {peer_id}...");
+
+                    let (task, handle) = self.message_sender();
+                    let future = self.runtime.spawn(task.exec());
+                    drop(future); // drop `JoinFuture` explicitly to make clippy happy
+
+                    self.senders.insert(peer_id, handle);
+
+                    self.senders
+                        .get(&peer_id)
+                        .expect("message sender task spawned")
+                }
+            };
+
+            match sender.try_send(message) {
+                Ok(()) => (),
+
+                Err(tokio::sync::mpsc::error::TrySendError::Full(message)) => {
+                    if log::max_level() >= log::Level::Debug {
+                        log::error!(
+                            "Failed to forward message {message:?} to message sender task {peer_id}: \
+                             message sender task queue is full"
+                        );
+                    } else {
+                        log::error!(
+                            "Failed to forward message to message sender task {peer_id}: \
+                             message sender task queue is full"
+                        );
+                    }
+                }
+
+                // This should *never* happen...
+                Err(tokio::sync::mpsc::error::TrySendError::Closed(message)) => {
+                    unreachable!(
```

Comment:
Sorry for focusing on that line.

I am thinking of an alternative because things that are supposed to never happen tend to happen :smile: 

Instead of exploding with a panic/unreachable, I propose to:
- log a error to help us understand the issue
- remove the sender for that peer `self.senders.remove(&peer_id)` as it reached an invalid state

This way we give a chance to the system to heal automatically with the next message for that peer.

WDYT?

Reply:
> I am thinking of an alternative because things that are supposed to never happen tend to happen

very much agree here

Reply:
Added a log and some retry handling (though, this may be a bit too much, let me know what you think).